### PR TITLE
Support Apple Silicon

### PR DESCRIPTION
**Pull request by JoshuaKirby88**
## Problem

The project didn’t support running on macOS Apple Silicon (MPS).

## Approach

- Added a macOS Apple Silicon setup guide (`MACOS_SETUP.md`) and linked it from the `README.md`.
- Updated GPU detection / GPU worker selection to recognize MPS in addition to CUDA, while keeping the existing CUDA path as-is as much as possible (and only adding a safe MPS check on the non-CUDA/CPU path).
- Updated the agent’s device-selection guidance and seed-setting snippet to handle MPS.
- Updated the `i_cant_believe_its_not_better*` examples to select `cuda` → `mps` → `cpu`, and guarded `torch.compile` / MPS checks for compatibility.

## Limitations

- MPS is treated as a single device, and multi-device scheduling is still CUDA-oriented.
- This improves defaults and common entrypoints, but LLM-generated experiment code can still explicitly reference `cuda`.
- No tests are added for macOS/MPS behavior.

### CODE REVIEW

File: ai_scientist/treesearch/parallel_agent.py [hunk 1]
```diff
@@ -1283,7 +1298,7 @@ def _run_multi_seed_evaluation(self, node: Node) -> List[Node]:
 
             # Add seed to node code
             node_data["code"] = (
-                f"# Set random seed\nimport random\nimport numpy as np\nimport torch\n\nseed = {seed}\nrandom.seed(seed)\nnp.random.seed(seed)\ntorch.manual_seed(seed)\nif torch.cuda.is_available():\n    torch.cuda.manual_seed(seed)\n\n"
+                f"# Set random seed\nimport random\nimport numpy as np\nimport torch\n\nseed = {seed}\nrandom.seed(seed)\nnp.random.seed(seed)\ntorch.manual_seed(seed)\nif torch.cuda.is_available():\n    torch.cuda.manual_seed(seed)\nelif torch.backends.mps.is_available():\n    torch.mps.manual_seed(seed)\n\n"
```
[the remark this answers is not in the document]
Comment (JoshuaKirby88) [reply ai_scientist/treesearch/parallel_agent.py]: This is inaccurate for newer PyTorch versions.

File: ai_scientist/ideas/i_cant_believe_its_not_better.py [hunk 2]
```diff
@@ -225,7 +231,7 @@ def __getitem__(self, idx):
 
 start_time = time.time()
 # Use torch.compile with safer settings
-if torch.cuda.is_available():
+if torch.cuda.is_available() or torch.backends.mps.is_available():
```
[the remark this answers is not in the document]
Comment (JoshuaKirby88) [reply ai_scientist/ideas/i_cant_believe_its_not_better.py]: Fixed in e216d76.

File: ai_scientist/ideas/i_cant_believe_its_not_betterrealworld.py [hunk 3]
```diff
@@ -193,7 +199,7 @@ def __getitem__(self, idx):
 
 start_time = time.time()
 # Use torch.compile with safer settings
-if torch.cuda.is_available():
+if torch.cuda.is_available() or torch.backends.mps.is_available():
```
[the remark this answers is not in the document]
Comment (JoshuaKirby88) [reply ai_scientist/ideas/i_cant_believe_its_not_betterrealworld.py]: Fixed in e216d76.

File: ai_scientist/treesearch/parallel_agent.py [hunk 4]
```diff
@@ -1135,7 +1140,17 @@ def get_gpu_count() -> int:
...
+            if len(devices) > 0:
+                return len(devices)
+
+        # Check for Apple Silicon MPS with torch
+        try:
+            import torch
+            if torch.backends.mps.is_available():
+                return 1  # MPS presents as a single device
+        except Exception:
+            pass
```
[the remark this answers is not in the document]
Comment (JoshuaKirby88) [reply ai_scientist/treesearch/parallel_agent.py]: Fixed in 725b087.

**Comment by Hello-GregKulp** [comment]
this would be great ~ thank you @JoshuaKirby88

**Comment by only976** [comment]
Hope this branch could merge to main soon.

**Comment by TomLucidor** [comment]
Related? https://github.com/SakanaAI/AI-Scientist-v2/pull/80

**Comment by alex-pythonista** [comment]
This is helpful. Hoping to get merged soon.
